### PR TITLE
Polish derviations

### PR DESCRIPTION
**Pull request by Quafadas**
To be clear, I don't think you should actually merge this!

Attached are some naive implementations of other "fundamental" things that may make sense to "auto derive". I think really I'm proposing a separate branch with a "polished" version of the concept you outlined in the tutorial? 

Note, that "polished" branch may not need to include the rough and ready implementations in this request!

### CODE REVIEW

**Comment by kitlangton** [comment]
Very awesome stuff :)

It would be great to polish this “library” up and make it more than just an example. Some other areas to think about are validations, not requiring the user to provide a default value, input masks, custom styles. Happy to pair on that sometime if you’re interested.

File: src/main/scala/formula/Formula.scala [hunk 1]
```diff
@@ -68,20 +61,74 @@ object Form {
...
+      )    
+  }
+
+  //The double implementation below is kind of nasty. But I think it works. Could be improved with a better error message perhaps.
+  implicit val dble: Form[Double] = new Form[Double] {
+    override def renderImpl(variable: Var[Double])(implicit owner: Owner): HtmlElement = {
+      def testFct(s:String) : Boolean = s.toDoubleOption match {
+              case Some(success) => {true}
+              case None => {false } 
+      }
```
Comment (kitlangton) [inline src/main/scala/formula/Formula.scala]: `s.toDoubleOption.isDefined` actually works as well :)

File: src/main/scala/formula/Formula.scala [hunk 2]
```diff
@@ -68,20 +61,74 @@ object Form {
...
+      val intermediate = variable.zoom[String](_.toString())(_.toDouble)
+      val fake = Var(intermediate.now())
+      
+      input(
+          value <-- intermediate,
+          onInput.mapToValue.filter({ina => 
+            fake.set(ina);
+            testFct(ina)
+          }) --> intermediate,
```
Comment (kitlangton) [inline src/main/scala/formula/Formula.scala:112]: Great idea :)

File: src/main/scala/formula/Formula.scala (hunk 1 above)
Comment (Quafadas) [reply src/main/scala/formula/Formula.scala]: :-)

**Comment by Quafadas** [comment]
I'd certainly be happy to pair up - but I warn you in advance I'm very much an amateur. I'm not even a professional software engineer... I think I'd get more out of it than you would! As long as you're comfortable with that... 

I'm happy to try and commit what time I can - the time I have at the moment I spend noodling around with scala and laminar  anyway -  but that time is often limited. 

So a qualified "yes" - I'd be happy to help, if I can, turn this into a 'library' (I could wheel into my own projects)...

On your initial thoughts; 
**not requiring the user to provide a default value**
A "blank string" is an invalid intermediate representation of most types... so I think this one could fit naturally in the concept above if we provide the UI user a good type hint and example (via placeholder perhaps?)

**validations**
Yes - was thinking that if the "testFunction" concept above can be bubbled up to library user land, then validations could follow that pattern?

 **input masks**
I don't honestly understand what this means :-(... ooffff...

 **custom styles**
I know I would want this, but I have nearly no idea how to do it in a robust manner in Laminar - learning curve too sharp - I currently rely on auld skool css most of the time.

**Comment by kitlangton** [comment]
Closing this, because I've started to work more on this, and have rewritten most of the internals :D But hopefully the upcoming improvements will address all these issues and more. :)

**Pull request closed without merging**